pkg/serviceaccount/metadata_test.go

```diff
@@ -0,0 +1,240 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package serviceaccount_test
+
+import (
+	"crypto/ecdsa"
+	"crypto/rsa"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"net/http/httptest"
+	"reflect"
+	"testing"
+
+	restful "github.com/emicklei/go-restful"
+	jose "gopkg.in/square/go-jose.v2"
+	"k8s.io/kubernetes/pkg/serviceaccount"
+)
+
+func setupServer(t *testing.T, iss string, keys []interface{}) *httptest.Server {
+	t.Helper()
+
+	c := restful.NewContainer()
+	s := httptest.NewServer(c)
+
+	// Install after we start server s.t. key URL can include host
+	srv, err := serviceaccount.NewServer(iss, s.URL, keys)
+	if err != nil {
+		t.Fatal(err)
+	}
+	srv.SetErrorHandler(func(err error) {
+		t.Fatal(err)
+	})
+	srv.Install(c)
+
+	return s
+}
+
+var defaultKeys = []interface{}{getPublicKey(rsaPublicKey), getPublicKey(ecdsaPublicKey)}
+
+// Configuration is an OIDC configuration, including all required fields.
+// https://openid.net/specs/openid-connect-discovery-1_0.html#ProviderMetadata
+type Configuration struct {
+	Issuer string `json:"issuer"`
+	// AuthzEndpoint string `json:"authorization_endpoint"` // REQUIRED, but useless to relying parties.
+	// TokenEndpoint `json:"token_endpoint"` // "REQUIRED unless only the Implicit Flow is used."
+	JwksURI       string   `json:"jwks_uri"`
+	ResponseTypes []string `json:"response_types_supported"`
+	SigningAlgs   []string `json:"id_token_signing_alg_values_supported"`
+	SubjectTypes  []string `json:"subject_types_supported"`
+}
+
+func TestServeConfiguration(t *testing.T) {
+	s := setupServer(t, "my-fake-issuer", defaultKeys)
+	defer s.Close()
+
+	want := Configuration{
+		Issuer:        "my-fake-issuer",
+		JwksURI:       s.URL + "/serviceaccountkeys/v1/jwks.json",
+		ResponseTypes: []string{"id_token"},
+		SubjectTypes:  []string{"public"},
+		SigningAlgs:   []string{"ES256", "RS256"},
+	}
+
+	reqURL := s.URL + "/.well-known/openid-configuration"
+
+	resp, err := http.Get(reqURL)
+	if resp == nil || err != nil {
```

Comment:
Don't check resp == nil, change the next t.Errorf to t.Fatalf, and defer resp.Body.Close() after this if.

Reply:
Done